test/e2e/hotGraphic.cy.js

```diff
@@ -0,0 +1,54 @@
+describe('Hot Graphic', function () {
+  function loopThroughHotGraphic(hotGraphicComponent) {
+    const itemsCount = Object.keys(hotGraphicComponent._items).length
+    cy.get('.hotgraphic__pin-item').should('have.length', itemsCount)
+    for (let i = 0; i < itemsCount; i++) {
+      cy.get(`.hotgraphic__pin-item .item-${i}.is-visited`).should('not.exist')
+      cy.get('.notify__popup.hotgraphic').should('not.exist')
+      cy.get(`.hotgraphic__pin-item .item-${i}`).click()
+      cy.get('.notify__popup.hotgraphic').should('be.visible')
+      cy.testContainsOrNotExists('.hotgraphic-popup__item-title-inner', hotGraphicComponent._items[i].title)
+      cy.testContainsOrNotExists('.hotgraphic-popup__item-body-inner', hotGraphicComponent._items[i].body)
+      cy.get('button.hotgraphic-popup__close').click()
+      cy.get(`.hotgraphic__pin-item .item-${i}.is-visited`).should('exist')
+    }
+
+    cy.get('.hotgraphic__pin-item .item-0').click()
+
+    for (let i = 0; i < itemsCount-1; i++) {
+      cy.testContainsOrNotExists('.hotgraphic-popup__item-title-inner', hotGraphicComponent._items[i].title)
+      cy.testContainsOrNotExists('.hotgraphic-popup__item-body-inner', hotGraphicComponent._items[i].body)
+      cy.get('.hotgraphic-popup__controls.next').click()
+    }
+    cy.get('.hotgraphic-popup__controls.next.is-disabled').should('exist')
+    for (let i = itemsCount-1; i >= 0; i--) {
+      cy.testContainsOrNotExists('.hotgraphic-popup__item-title-inner', hotGraphicComponent._items[i].title)
+      cy.testContainsOrNotExists('.hotgraphic-popup__item-body-inner', hotGraphicComponent._items[i].body)
+      cy.get('.hotgraphic-popup__controls.back').click()
+    }
```

Comment:
rather than looping through the items twice here to test the pin items you should be able to account for whether or not the elements with classes '.hotgraphic-popup__controls.next.is-disabled' and '.hotgraphic-popup__controls.back.is-disabled' exist in one loop, e.g. something like

```
cy.get('.hotgraphic-popup__controls.back.is-disabled').should('exist')
for (let i = 0; i < itemsCount-1; i++) {
      cy.testContainsOrNotExists('.hotgraphic-popup__item-title-inner', hotGraphicComponent._items[i].title)
      cy.testContainsOrNotExists('.hotgraphic-popup__item-body-inner', hotGraphicComponent._items[i].body)
      cy.get('.hotgraphic-popup__controls.next').click()
    }
    cy.get('.hotgraphic-popup__controls.next.is-disabled').should('exist')
    }
```


Also might be worth putting this into a separate test e.g. `it('Can cycle through each hotgraphic within the popup', () => {...` as it can was hard to tell at first glance what is being tested here.

Reply:
I'd like to keep them all in one test as it needs to loop through each component anyway, otherwise you're just looping over and over again. I would happily add comments to the tests to explain what's going on though, if other's would be happy with that

Reply:
comments would definitely help, or you could put the code into functions with appropriate names to make the code easier to read.